cedar-policy-core/src/entities/conformance.rs

```diff
@@ -176,6 +179,66 @@ pub fn typecheck_value_against_schematype(
     }
 }
 
+/// Check whether the given `RestrictedExpr` is a valid instance of `SchemaType`
+pub fn does_restricted_expr_implement_schematype(
+    expr: BorrowedRestrictedExpr<'_>,
+    expected_ty: &SchemaType,
+) -> bool {
+    use SchemaType::*;
+
+    match expected_ty {
+        Bool => expr.as_bool().is_some(),
+        Long => expr.as_long().is_some(),
+        String => expr.as_string().is_some(),
+        EmptySet => expr.as_set_elements().is_some(),
+        // PANIC SAFETY: LHS checks that `is_some()`.
+        #[allow(clippy::unwrap_used)]
+        Set { .. }
+            if expr.as_set_elements().is_some() && expr.as_set_elements().unwrap().count() == 0 =>
+        {
+            true
+        }
+        Set { element_ty: elty } => match expr.as_set_elements() {
+            Some(mut els) => els.all(|e| does_restricted_expr_implement_schematype(e, elty)),
+            None => false,
+        },
+        Record { attrs, open_attrs } => match expr.as_record_pairs() {
+            Some(pairs) => {
+                let pairs_map: BTreeMap<&SmolStr, BorrowedRestrictedExpr<'_>> = pairs.collect();
+                let all_req_schema_attrs_in_record = attrs.iter().all(|(k, v)| {
+                    !v.required
+                        || match pairs_map.get(k) {
+                            Some(inner_e) => {
+                                does_restricted_expr_implement_schematype(*inner_e, &v.attr_type)
+                            }
+                            None => false,
+                        }
+                });
+                let all_rec_attrs_in_schema =
+                    pairs_map.iter().all(|(k, inner_e)| match attrs.get(*k) {
+                        Some(sch_ty) => {
+                            does_restricted_expr_implement_schematype(*inner_e, &sch_ty.attr_type)
+                        }
+                        None => false,
+                    });
+                (*open_attrs || all_rec_attrs_in_schema) && all_req_schema_attrs_in_record
```

Comment:
Does this do the right thing in the case where we have `open_attrs == true`, there is an optional attribute, and `pairs_map` has a value for the optional attribute but it's not the type declared in the schema?

In my reading, `all_rec_attrs_in_schema` will be `false`, but we'll still pass overall due to the logic in line 224

Reply:
Good catch!

Reply:
Added two new tests